### PR TITLE
Introduce a collector manager for QueryPhaseCollector

### PR DESCRIPTION
**Pull request by javanna**
The query phase now relies on `QueryPhaseCollector` to wrap top docs and aggs collector as well as to apply `terminate_after`, `post_filter` and `min_score`. In order to enable inter-segment concurrency, we need to introduce a corresponding collector manager that can create multiple query phase collectors and reduce them to a single resultset.

Note: we are only introducing the collector manager and corresponding tests, but not using it yet in `QueryPhase`.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java [hunk 1]
```diff
@@ -102,6 +102,8 @@ public void testNegativeTerminateAfter() {
         );
     }
 
+    // TODO remove some of these tests as they are now duplicated?
+
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:106]: I wonder what your opinion is on test coverage: we are duplicating tests right now, is it still needed to test sequential execution explicitly that does not go through the manager? Some of these existing tests are very specific to the collector which I am going to keep, but the ones that I copied to the new test class and modified to use a collector manager could be removed? I am not entirely sure.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+class QueryPhaseCollectorManager<TD, AG> implements CollectorManager<QueryPhaseCollector, QueryPhaseCollectorManager.Result<TD, AG>> {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:20]: the types are slightly controversial right now. They ease testing for now. Especially the aggs now we are not going to leverage in practice outside of tests. I am thinking that the fact that they help testing is a good reason to keep them, but I am open to other ideas.

**Review by cbuescher — COMMENTED** [review]
I left a couple of smaller questions but nothing unexpected. I ran the new tests a bit more extensively and all the "terminateAfter" type of test fail once or twice out of a few hundred runs with an off-by-one error like

```
REPRODUCE WITH: ./gradlew 'null' --tests "org.elasticsearch.search.query.QueryPhaseCollectorManagerTests.testTerminateAfterTopDocsOnly [seed=[C610A5E08C531B52:E34FF37D0988682F]]" -Dtests.seed=C610A5E08C531B52 -Dtests.locale=el-CY -Dtests.timezone=America/Buenos_Aires -Druntime.java=17

java.lang.AssertionError: 
Expected :566
Actual   :567
<Click to see difference>
[...]
org.elasticsearch.search.query.QueryPhaseCollectorManagerTests.testTerminateAfterTopDocsOnly(QueryPhaseCollectorManagerTests.java:458)
```

Unfortunately the seed doesn't help me reproducing the error locally, but something to look into.

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java (hunk 1 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:106]: I didn't take a closer side-by-side look yet, but any methods you are considering specifically?

**Comment by javanna** [comment]
@cbuescher I am aware of the failure, I've seen it too and I will look into addressing it before merging. It looks to be a timing issue because it does not repro with the seed?

**Comment by cbuescher** [comment]
> looks to be a timing issue because it does not repro with the seed

That's what I thought, also interesting that it seems to be almost always off by one for all errors I've seen so far (across all "terminateAfterXYZ" methods)

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java (hunk 1 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:106]: > the ones that I copied to the new test class and modified to use a collector manager could be removed?

**Comment by javanna** [comment]
> That's what I thought, also interesting that it seems to be almost always off by one for all errors I've seen so far (across all "terminateAfterXYZ" methods)

exactly, all of the terminate after tests may fail with an off by one error, that is sneaky. I will dig into it.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java [hunk 3]
```diff
@@ -334,4 +349,61 @@ public DocIdSetIterator competitiveIterator() throws IOException {
...
         }
     }
+
+    static <T, A> QueryPhaseCollectorManager<T, A> createManager(
+        CollectorManager<? extends Collector, T> topDocsCollectorManager,
+        Weight postFilterWeight,
+        int terminateAfter,
+        CollectorManager<? extends Collector, A> aggsCollectorManager,
+        Float minScore
+    ) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java:359]: I am struggling here with all these generics, are they really necessary?
 topDocsCollectorManager and aggsCollectorManager don't need to be lucene interfaces. They are not used at that level, why are we trying to use them? makes the code so fragile, with all theses generics and castings on the reduce method. 
Please @javanna, reconsider an approach that does not use generic, we don't really need to go this path.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java (hunk 2 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:20]: Do we really need to return that Result object? I prefer tp keep doing the same thing as we were doing before and register results on the query result object instead of trying to return the values in such a way. 

Sorry again, but I struggle to see the benefits of generics here.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:20]: I am hoping that you'll see the benefits when I am done integrating it in the query phase, which this PR does not do.

Bear with me, a couple more PRs are needed to complete this work and the types are so far used to simplify testing and retrieve the results from the two collector managers. That is why I left myself a question asking for feedback. I can look into removing the type for aggs, as it is not strictly needed, although not having it will complicate a bit retrieving the aggs result in the test.

The long term plan is to remove the indirections / supplier in TopDocsCollectorManagerFactory: based on the collector chosen, we build total hits, and top docs differently. All that can be integrated into the query phase collector manager, so that the search method returns the actual thing that gets set to the query result. Right now, we set the top docs to the query result but it is hard to follow where the top docs come from, as it is done through the suppliers. I think that this will make things better compared to how they currently are.

That is my reasoning for why we may not need generics for aggs, but we will use them at our advantage for aggs. Plus I don't see how not to deal with generics  entirely when we rely on Lucene design that does have a certain contract / constraints which revolves around the reduce method.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:20]: > the same thing as we were doing before

To clarify, what we had in previous prototypes required saving the results of the reduction in the collector manager and adding methods to retrieve it, while generics allow us to retrieve the results directly from the search method. This may not work with aggs, as I understand they may not rely on calling reduce but for top docs collection, it is beneficial.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java (hunk 3 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java:359]: > topDocsCollectorManager and aggsCollectorManager don't need to be lucene interfaces. They are not used at that level, why are we trying to use them?

they currently are both collector managers in the query phase, that create one collector per slice. That is at least the case for top docs collection. Can you expand on what you are suggesting.

> makes the code so fragile, with all theses generics and castings on the reduce method.

The casting is necessary because `List<Collector>` is different from `List<? extends Collector>` which is extremely odd. I don't see this as a blocker though as long as that's the only place where casting happens, for which we suppress warnings at the line level. I failed to see how this makes the code fragile. Like I said above, I will look into removing the type for aggs as it seems like we won't need in practice and it's odd to need it only for tests.

Hopefully, things will be cleaner and clearer as I can make further changes towards the long term direction that I mentioned above.

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+            );
+            QueryPhaseCollectorManager.Result<TopDocs, Integer> result = searcher.search(new MatchAllDocsQuery(), manager);
+            assertFalse(result.terminatedAfter());
+            assertEquals(numDocs, result.topDocs().totalHits.value);
+            assertEquals(numDocs, result.aggs().intValue());
+        }
+        {
+            CollectorManager<TopScoreDocCollector, TopDocs> topDocsManager = TopScoreDocCollector.createSharedManager(1, null, 1000);
+            CollectorManager<DummyTotalHitCountCollector, Integer> aggsManager = DummyTotalHitCountCollector.createManager();
+            TermQuery termQuery = new TermQuery(new Term("field2", "value"));
```
Comment (iverase) [inline server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java:178]: Let's don't use dummy collectors where we should be using an aggregation specific collector, please. This will be limiting in the long term if we want to do some more specific aggregation stuff here.

Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java:178]: Is there a lightweight aggregation specific collector that I can use? Can you point me to how I can create one please?

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+    private Directory directory;
+    private IndexReader reader;
+    private IndexSearcher searcher;
+    private int numDocs;
+    private int numField2Docs;
+    private int numField3Docs;
+    private int numField2AndField3Docs;
+
+    @Override
+    public void setUp() throws Exception {
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java:50]: will do so

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+        public Void reduce(Collection<Collector> collectors) throws IOException {
+            @SuppressWarnings("unchecked")
+            List<C> cs = collectors.stream().map(collector -> (C) collector).toList();
+            result = wrapped.reduce(cs);
+            return null;
+        }
+
+        public T getResult() {
+            return result;
+        }
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java:112]: I relegated the typing and awkward things in this test class for now. This is what it takes to be able to get results of the collection if we don't leverage the return type of the search method. Less generics, but each collector manager needs to hold mutable state that is assigned when reduce is called, and additional getters to retrieve them. In the long run, the query phase will have a well defined set of expected return types that can simplify things.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java (hunk 3 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java:348]: I removed the generics for now. I will reintroduce them back later when they make more sense in the query phase. This does mean that getting results out of the collector managers requires wrapping in this test, to allow  holding the result of the reduce call which could rather be returned directly from the `search` method.

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java (hunk 1 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:140]: I have moved over all the search tests to using the collector manager. The collector low-level tests that don't do a search still test the collector directly.

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java (hunk 4 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java]: I am leaning towards changing this as a follow-up. I would gladly take help on this. I tried it and probably took a wrong direction but it required a non negligible amount of context related objects.

By the way I have attempted to do the same (and failed so far) in `QueryPhaseTests`, as we have many variations of tests there but never aggs in the mix. I would say that if we want real aggs collectors `QueryPhaseTests` would be a good place to start, as it tests the interaction between the different parts of the query phase, while this specific test is testing a specific collector, which I think makes it more suited for dummy/simple wrapped collectors. That said if there is a way to get something more similar to what we have in real-life, which does not have too many dependencies, that seems like a good addition to this test too.

**Comment by javanna** [comment]
@cbuescher the change to the initialization of the test makes the failures you've seen harder to reproduce, but I made sure to address that before changing the test. We were not incrementing and checking the threshold atomically. It is debatable whether that is needed, given the potential overhead. We could accept some discrepancies when going concurrent, which though makes testing more difficult. I would opt for precision to start with, and unifying the behaviour between sequential and concurrent execution. The important bit is that there is overhead only when terminate_after is set, and we expect `terminate_after` to be used with relatively low values, where performance does not matter as much as the actual early termination.

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java (hunk 4 above)
Comment (iverase) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java]: Agree, there is not an easy solution as far as we make aggregations look like lucene collectors.

**Review by cbuescher — APPROVED** [review]
Sorry for the delay getting back to this. Since the collector manager that I commented on last has moved out to another PR I only looked at the incremental changes since then, all LGTM
